python/cog/server/worker.py

```diff
@@ -44,44 +46,164 @@
+    #         if self._child.is_alive():
+    #             self._child.terminate()
+    #             self._child.join()
 
     def setup(self) -> AsyncIterator[_PublicEventType]:
         self._assert_state(WorkerState.NEW)
         self._state = WorkerState.STARTING
-        self._child.start()
 
-        return self._wait(raise_on_error="Predictor errored during setup")
+        async def inner() -> AsyncIterator[_PublicEventType]:
+            # in 3.10 Event started doing get_running_loop
+            # previously it stored the loop when created, which causes an error in tests
+            if sys.version_info < (3, 10):
+                self._terminating = self._mux.terminating = asyncio.Event()
+
+            self._child.start()
+            self._ensure_event_reader()
+            async for event in self._mux.read("SETUP", poll=0.1):
+                yield event
+                if isinstance(event, Done):
+                    if event.error:
+                        raise FatalWorkerException(
+                            "Predictor errored during setup: " + event.error_detail
+                        )
+                    self._state = WorkerState.IDLE
+
+        return inner()
+
+    @contextlib.asynccontextmanager
+    async def prediction_ctx(self, input: PredictionInput) -> AsyncIterator[None]:
+        async with self._semaphore:
+            if self._semaphore._value == 0:
+                # maybe this will fix hypothesis
+                self._state = WorkerState.BUSY
+            self._allow_cancel = True
+            self._predictions_in_flight.add(input.id)
+            try:
+                yield
+            finally:
+                self._predictions_in_flight.remove(input.id)
+        if self._semaphore._value == self._concurrent:
+            self._state = WorkerState.IDLE
+            self._allow_cancel = False
+        else:
+            # we just finished a prediction, so if we were BUSY we aren't anymore
+            self._state = WorkerState.PROCESSING
 
     def predict(
         self, payload: Dict[str, Any], poll: Optional[float] = None
     ) -> AsyncIterator[_PublicEventType]:
-        self._assert_state(WorkerState.READY)
+        # this has to be eager just for hypothesis
+        if self._state not in {WorkerState.PROCESSING, WorkerState.IDLE}:
+            raise InvalidStateException(
+                f"Invalid operation: state is {self._state} (must be processing or idle)"
+            )
+        if self._shutting_down:
+            raise InvalidStateException(
+                "cannot accept new predictions because shutdown requested"
+            )
         self._state = WorkerState.PROCESSING
-        self._allow_cancel = True
-        self._events.send(PredictionInput(payload=payload))
 
-        return self._wait(poll=poll)
+        async def inner() -> AsyncIterator[_PublicEventType]:
+            input = PredictionInput(payload=payload)
+            async with self.prediction_ctx(input):
```

Comment:
@nickstenning, supporting hypothesis like this kind of ruins the semantics of the semaphore. do you have any suggestions?

Reply:
We've chatted about this since you left this comment, but just to reiterate, I don't care about "supporting Hypothesis," only about verifying (with Hypothesis or otherwise) that expected things happen if you call the various public methods of `Worker` at any point.

I'm not quite sure what you mean about ruining the semantics of the semaphore? If you want a preemptive check of the state of the semaphore then `sem.locked()` exists, but also as this part of the code is not actually multithreaded I do wonder whether a plain old integer on `Worker` would also work just fine.

Reply:
I've figured this out to my satisfaction in this PR. #1499 addresses this in further depth and clarifies the distinction between predictions_in_flight (updated as soon as work is accepted) and the semaphore (updated once the work is started). the semaphore is the only part that makes it an async context manager; a sync context manager is sufficient for the eager part of the state management